src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterDownload.java

```diff
@@ -178,170 +185,441 @@ public Resource execute(String mimeType, final ProgressListener progressListener
             }
-            if (clippedGridCoverage != null) {
-                resourceManager.addResource(new GridCoverageResource(clippedGridCoverage));
+
+            // Reproject if needed
+            if (reproject) {
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.log(Level.FINE, "Reprojecting the coverage");
+                }
+                disposableSources.add(gridCoverage);
+                gridCoverage = (GridCoverage2D) Operations.DEFAULT.resample(
+                        gridCoverage, targetCRS, null, interpolation, backgroundValues);
             }
-            if (scaledGridCoverage != null) {
-                resourceManager.addResource(new GridCoverageResource(scaledGridCoverage));
+            return gridCoverage;
+        }
+
+    }
+
+    /**
+     * Extract the backgroundValues which will be used for mosaicking operations.
+     *  
+     * @param coverageParameters
+     * @param readParameters
+     * @return
+     */
+    private double[] getBackgroundValues(Map<String, Serializable> coverageParameters,
+            GeneralParameterValue[] readParameters) {
+        double[] backgroundValues = null;
+        if (coverageParameters != null && coverageParameters.containsKey("BackgroundValues")) {
+            for (GeneralParameterValue readParameter : readParameters) {
+                if ("BackgroundValues"
+                        .equalsIgnoreCase(readParameter.getDescriptor().getName().toString())) {
+                    Object bgValue = ((ParameterValue) readParameter).getValue();
+                    if (bgValue != null && bgValue instanceof double[]) {
+                        backgroundValues = ((double[]) bgValue);
+                    }
+                    break;
+                }
             }
         }
+        return backgroundValues;
+    }
+
+    /**
+     * Check if a crop operation is needed (meaning that the extent of the returned image contains 
+     * the requested extent).
+     * 
+     * @param rasterData
+     * @param requestedRange
+     * @return
+     */
+    private boolean cropIsNeeded(RenderedImage rasterData, GridEnvelope requestedRange) {
+        final int requestedW = requestedRange.getSpan(0);
+        final int requestedH = requestedRange.getSpan(1);
+        final int requestedMinX = requestedRange.getLow(0);
+        final int requestedMinY = requestedRange.getLow(1);
+        final int requestedMaxX = requestedRange.getHigh(0);
+        final int requestedMaxY = requestedRange.getHigh(1);
+
+        final int imageW = rasterData.getWidth();
+        final int imageH = rasterData.getHeight();
+        final int minX = rasterData.getMinX();
+        final int minY = rasterData.getMinY();
+        final int maxX = minX + imageW - 1;
+        final int maxY = minY + imageH - 1;
+
+        if (((imageW + ((minX >= 0) ? 0 : minX)) >= requestedW)
```

Comment:
I don't get while the min values are summed to the width, shouldn't it be a simple width vs width comparison?
I have the impression that building two java.awt.Rectangle and doing a contains test would probably be shorter to write and easier to understand (question, can there be the case where an image is outside the requested area? or one where the image is just translated vs what was requested?)

Reply:
This is a check to deal with some corner cases occurred in a different project using wps.
I would leave it as is.

Reply:
Ok